File: modules/sentiment-analysis/sentiment-analysis-score-data.r
#' Sentiment Analysis with Dictionary Analysis & BERT

# Initialize ----------------------------------------------------------

## Set Constants ----------------------------------------------------------
JOB_NAME = 'sentiment-analysis-score-data'
EF_DIR = Sys.getenv('EF_DIR')
RESET_SQL = FALSE
MAX_ROWS = 50000 # Max numbers of rows to process with jobs

## Cron Log ----------------------------------------------------------
if (interactive() == FALSE && rstudioapi::isAvailable(child_ok = T) == F) {
	sink_path = file.path(EF_DIR, 'logs', paste0(JOB_NAME, '.log'))
	sink_conn = file(sink_path, open = 'at')
	system(paste0('echo "$(tail -50 ', sink_path, ')" > ', sink_path,''))
	lapply(c('output', 'message'), function(x) sink(sink_conn, append = T, type = x))
	message(paste0('\n\n----------- START ', format(Sys.time(), '%m/%d/%Y %I:%M %p ----------\n')))
}

## Load Libs ----------------------------------------------------------'
library(econforecasting)
library(tidyverse)
library(data.table)
library(DBI)
library(lubridate)
library(tidytext)
library(reticulate)
use_virtualenv(file.path(EF_DIR, '.virtualenvs', 'econforecasting'))

## Load Connection Info ----------------------------------------------------------
db = connect_db(secrets_path = file.path(EF_DIR, 'model-inputs', 'constants.yaml'))
run_id = log_start_in_db(db, JOB_NAME, 'sentiment-analysis')


# Data Import --------------------------------------------------------

## Create Tables --------------------------------------------------------
local({
if (RESET_SQL) {

	dbExecute(db, 'DROP TABLE IF EXISTS sentiment_analysis_reddit_score CASCADE')
	dbExecute(db, 'DROP TABLE IF EXISTS sentiment_analysis_media_score CASCADE')

	dbExecute(
		db,
		'CREATE TABLE sentiment_analysis_reddit_score (
		scrape_id INT NOT NULL,
		text_part VARCHAR(255) NOT NULL,
		score_model VARCHAR(255) NOT NULL,
		score VARCHAR(100) NOT NULL,
		score_conf DECIMAL(20, 4) NULL,
		scored_dttm TIMESTAMP WITH TIME ZONE NOT NULL,
		PRIMARY KEY (scrape_id, text_part, score_model),
		CONSTRAINT sentiment_analysis_reddit_score_fk FOREIGN KEY (scrape_id)
				REFERENCES sentiment_analysis_reddit_scrape (id) ON DELETE CASCADE ON UPDATE CASCADE
		)'
	)

	dbExecute(
		db,
		'CREATE TABLE sentiment_analysis_media_score (
		scrape_id INT NOT NULL,
		text_part VARCHAR(255) NOT NULL,
		score_model VARCHAR(255) NOT NULL,
		score VARCHAR(100) NOT NULL,
		score_conf DECIMAL(20, 4) NULL,
		scored_dttm TIMESTAMP WITH TIME ZONE NOT NULL,
		PRIMARY KEY (scrape_id, text_part, score_model),
		CONSTRAINT sentiment_analysis_media_score_fk FOREIGN KEY (scrape_id)
				REFERENCES sentiment_analysis_media_scrape (id) ON DELETE CASCADE ON UPDATE CASCADE
		)'
	)

}
})

## Boards --------------------------------------------------------
local({

	score_boards = collect(tbl(db, sql(
		"SELECT subreddit, scrape_ups_floor FROM sentiment_analysis_reddit_boards
		WHERE score_active = TRUE"
	)))

	score_boards <<- score_boards
})

## Pull Unscored Data --------------------------------------------------------
local({

	score_models = c('DICT', 'DISTILBERT', 'ROBERTA')

	unscored_text = lapply(score_models %>% setNames(., .), function(x)
		dbGetQuery(db, str_glue(
			"(
				SELECT
					'reddit' AS source, DATE(r1.created_dttm AT TIME ZONE 'US/Eastern') as created_dt,
					r1.id,
					CONCAT(r1.title, COALESCE(CONCAT('. ', r1.selftext), '')) AS text_part_all_text
				FROM sentiment_analysis_reddit_scrape r1
				LEFT JOIN
					(
					SELECT scrape_id--, DATE(scraped_dttm) AS scrape_date
					FROM sentiment_analysis_reddit_score
					WHERE score_model = '{x}'
					) r2
					ON r1.id = r2.scrape_id
				-- Only keep subreddits in scrape_boards directory and above score_ups_floor
				INNER JOIN sentiment_analysis_reddit_boards b
					ON r1.subreddit = b.subreddit AND b.scrape_active = TRUE AND r1.ups >= b.score_ups_floor
				WHERE r2.scrape_id IS NULL
					-- Rescore comments scored before 5/10/22
					--OR DATE(r2.scrape_date) <= '2022-05-01'
			)
			UNION ALL
			(
				SELECT
					'media' AS source, m1.created_dt,
					m1.id,
					CONCAT(m1.title, COALESCE(CONCAT('. ', m1.description), '')) AS text_part_all_text
				FROM sentiment_analysis_media_scrape m1
				LEFT JOIN
					(
					SELECT scrape_id--, DATE(scraped_dttm) AS scrape_date
					FROM sentiment_analysis_media_score
					WHERE score_model = '{x}'
					) m2
					ON m1.id = m2.scrape_id
				WHERE m2.scrape_id IS NULL
					-- Rescore comments scored before 5/10/22
					-- OR DATE(r2.scrape_date) <= '2022-05-01'
			)"
			)) %>%
			as_tibble(.) %>%
			mutate(
				.,
				text_part_all_text =
					text_part_all_text %>%
					str_replace_all(., '[^\x01-\x7F]', '') %>%
					str_squish(.) %>%
					str_sub(., 1, 512)
				)
	)

	unscored_text %>%
		bind_rows(.) %>%
		sample_n(., size = 20) %>%
		.$text_part_all_text %>%
		print(.)

	unscored_text <<- unscored_text
})


# Analysis --------------------------------------------------------

## Dictionary Analysis --------------------------------------------------------
local({

	BATCH_SIZE = 10000

	dict =
		list(
			get_sentiments('nrc') %>%
				as.data.table(.) %>%
				.[, sentiment := fcase(
					sentiment %chin% c('trust', 'positive', 'joy', 'anticipation', 'surprise'), 1,
					sentiment %chin% c('disgust', 'sadness', 'fear', 'negative', 'anger'), -1
				)],
			get_sentiments('bing') %>%
				as.data.table(.) %>%
				.[, sentiment := ifelse(sentiment == 'positive', 1, -1)]
		) %>%
		rbindlist(.) %>%
		.[, list(sentiment = head(sentiment, 1)), by = 'word']

	batches =
		unscored_text$DICT %>%
		as.data.table(.) %>%
		.[, split := ceiling((1:nrow(.))/BATCH_SIZE)] %>%
		split(., by = c('source', 'split'), keep.by = T, flatten = T) %>%
		unname(.)

	iwalk(batches, function(x, i) {

		message('***** Dictionary Scoring Batch ', i, ' of ', length(batches))

		result =
			x %>%
			.[, c('id', 'text_part_all_text')] %>% # Name is the unique post identifier
			tidytext::unnest_tokens(., word, text_part_all_text) %>%
			# .[!word %chin% stop_words$word]  %>%
			merge(., dict, by = 'word', all = F, allow.cartesian = T) %>%
			.[, list(score = sum(sentiment)/.N), by = 'id'] %>%
			.[, score := ifelse(score >= 0, 'p', 'n')]

		# print(result)
		if (nrow(result) == 0) {
			message('***** No dictionary matches, skipping')
			return()
		}

		sql_data =
			result %>%
			transmute(
				.,
				scrape_id = id,
				text_part = 'all_text',
				score_model = 'DICT',
				score,
				score_conf = NULL,
				scored_dttm = now()
				) %>%
			mutate(., across(where(is.POSIXt), function(x) format(x, '%Y-%m-%d %H:%M:%S %Z')))

		sql_data %>%
			create_insert_query(
				.,
				paste0('sentiment_analysis_', x$source[[1]], '_score'),
				'ON CONFLICT (scrape_id, text_part, score_model) DO UPDATE SET
					score=EXCLUDED.score,
					score_conf=EXCLUDED.score_conf,
					scored_dttm=EXCLUDED.scored_dttm'
				) %>%
			dbExecute(db, .)
	})

})

## DISTILBERT  --------------------------------------------------------
local({

	BATCH_SIZE = 1000
	message('UNSCORED DISTILBERT: ', nrow(unscored_text$DISTILBERT))

	happytransformer = import('happytransformer')
	happy_tc = happytransformer$HappyTextClassification(
		'DISTILBERT', # 'ROBERTA',
		'distilbert-base-uncased-finetuned-sst-2-english', # 'siebert/sentiment-roberta-large-english',
		num_labels = 2
	)

	batches =
		unscored_text$DISTILBERT %>%
			# Score first 10k only to prevent overlap
			head(., MAX_ROWS) %>%
			as.data.table(.) %>%
			.[order(-source, -created_dt)] %>%
			.[, split := ceiling((1:nrow(.))/BATCH_SIZE)] %>%
			split(., by = c('source', 'split'), keep.by = T, flatten = T) %>%
			unname(.)

	iwalk(batches, function(x, i) {

		pull_counts =
			x %>%
			.[, list(count = .N), by = 'created_dt'] %>%
			.[, x:= paste0(format(created_dt, "%m/%d/%y"), " (", count, ")")] %>%
			.$x %>%
			paste0(., collapse = "\n")

		message(str_glue('***** DISTILBERT Scoring {i} of {length(batches)} | Source: {x$source[[1]]} | Counts: \n{pull_counts}'))

		fwrite(set_names(x[, 'text_part_all_text'], 'text'), file.path(tempdir(), 'text.csv'))
		classified_text = happy_tc$test(file.path(tempdir(), 'text.csv'))

		result = data.table(
			scrape_id = x$id,
			score = map_chr(classified_text, ~ .$label),
			score_conf = map_chr(classified_text, ~ .$score)
			)[, score := ifelse(score == 'POSITIVE', 'p', 'n')]

		# print(result)

		sql_data =
			result %>%
			transmute(
				.,
				scrape_id,
				text_part = 'all_text',
				score_model = 'DISTILBERT',
				score,
				score_conf,
				scored_dttm = now()
				) %>%
			mutate(., across(where(is.POSIXt), function(x) format(x, '%Y-%m-%d %H:%M:%S %Z')))

		sql_data %>%
			create_insert_query(
				.,
				paste0('sentiment_analysis_', x$source[[1]], '_score'),
				'ON CONFLICT (scrape_id, text_part, score_model) DO UPDATE SET
					score=EXCLUDED.score,
					score_conf=EXCLUDED.score_conf,
					scored_dttm=EXCLUDED.scored_dttm'
			) %>%
			dbExecute(db, .)
	})

	# Parallelization
	# USE_CORES = 2
	#
	# batches =
	# 	unscored_bert %>%
	# 	# Score first X rows only to prevent overlap
	# 	head(., MAX_ROWS) %>%
	# 	as.data.table(.) %>%
	# 	.[, text_part_all_text := str_sub(text_part_all_text, 1, 512)] %>%
	# 	.[order(-source, -created_dt)] %>%
	# 	.[, split_core := ceiling((1:nrow(.))/(MAX_ROWS/USE_CORES))] %>%
	# 	split(., by = 'split_core', keep.by = F) %>%
	# 	lapply(., function(x) {
	# 		x %>%
	# 			.[, split := ceiling((1:nrow(.))/BATCH_SIZE)] %>%
	# 			split(., by = c('source', 'split'), keep.by = T, flatten = T) %>%
	# 			unname(.)
	# 	})
	#
	# ## Setup empty progress-bar file
	# progress_file = tempfile()
	# cat('', file = progress_file, append = F)
	#
	# library(future)
	# plan('multisession', workers = USE_CORES)
	#
	# ## Detect non-exportable objects and give an error asap
	# options(future.globals.onReference = "error")
	# promises = lapply(1:USE_CORES, function(i)
	# 	future({
	# 		pid = Sys.getpid()
	# 		# cat(str_glue('\nCore {i}/{pid} initialized'), file = progress_file, append = T)
	# 		message(pid)
	# 		library(reticulate)
	# 		# cat(str_glue('\nImporting Python Modules: {i}/{pid}'), file = progress_file, append = T)
	# 		import_path = file.path(EF_DIR, '.virtualenvs', 'econforecasting')
	# 		use_virtualenv(import_path)
	# 		happytransformer = reticulate::import_from_path(
	# 			'happytransformer',
	# 			path = file.path(EF_DIR, '.virtualenvs', 'econforecasting', 'lib64/python3.8/site-packages/happytransformer')
	# 			)
	# 		# happytransformer = reticulate::import('happytransformer')
	# 		# happy_tc = happytransformer$HappyTextClassification(
	# 		# 	'DISTILBERT', # 'ROBERTA',
	# 		# 	'distilbert-base-uncased-finetuned-sst-2-english', # 'siebert/sentiment-roberta-large-english',
	# 		# 	num_labels = 2
	# 		# )
	#
	# 		Sys.sleep(10)
	# 	}, globals = TRUE)
	# )
	#
	# ## Poll progress every 0.1 s until done
	# while (all(resolved(promises)) == FALSE) {
	# 	## Report on progress
	# 	print('test')
	# 	# cat(read_file(progress_file))
	# 	readLines(progress_file)
	# 	Sys.sleep(1)
	# }
	# value(promises[[1]])
})


## ROBERTA  --------------------------------------------------------
local({

	BATCH_SIZE = 1000
	message('UNSCORED ROBERTA: ', nrow(unscored_text$ROBERTA))

	happytransformer = import('happytransformer')
	happy_tc = 	happytransformer$HappyTextClassification(
		'DistilRoBERTa-base', # 'ROBERTA',
		'j-hartmann/emotion-english-distilroberta-base',
		num_labels = 7
	)

	batches =
		unscored_text$ROBERTA %>%
		head(., MAX_ROWS) %>%
		as.data.table(.) %>%
		.[order(-source, -created_dt)] %>%
		.[, split := ceiling((1:nrow(.))/BATCH_SIZE)] %>%
		split(., by = c('source', 'split'), keep.by = T, flatten = T) %>%
		unname(.)

	iwalk(batches, function(x, i) {

		pull_counts =
			x %>%
			.[, list(count = .N), by = 'created_dt'] %>%
			.[, x:= paste0(format(created_dt, "%m/%d/%y"), " (", count, ")")] %>%
			.$x %>%
			paste0(., collapse = "\n")

		message(str_glue('***** ROBERTA Scoring {i} of {length(batches)} | Source: {x$source[[1]]} | Counts: \n{pull_counts}'))

		fwrite(set_names(x[, 'text_part_all_text'], 'text'), file.path(tempdir(), 'text.csv'))
		classified_text = happy_tc$test(file.path(tempdir(), 'text.csv'))

		result = data.table(
			scrape_id = x$id,
			score = map_chr(classified_text, ~ .$label),
			score_conf = map_chr(classified_text, ~ .$score)
		)

		sql_data =
			result %>%
			transmute(
				.,
				scrape_id,
				text_part = 'all_text',
				score_model = 'ROBERTA',
				score,
				score_conf,
				scored_dttm = now()
			) %>%
			mutate(., across(where(is.POSIXt), function(x) format(x, '%Y-%m-%d %H:%M:%S %Z')))

		sql_data %>%
			create_insert_query(
				.,
				paste0('sentiment_analysis_', x$source[[1]], '_score'),
				'ON CONFLICT (scrape_id, text_part, score_model) DO UPDATE SET
			score=EXCLUDED.score,
			score_conf=EXCLUDED.score_conf,
			scored_dttm=EXCLUDED.scored_dttm'
			) %>%
			dbExecute(db, .)
	})

})

# Finalize --------------------------------------------------------

## Log --------------------------------------------------------
local({

	log_data = list(
		scored_dict = nrow(unscored_text$DICT),
		scored_roberta = nrow(unscored_text$ROBERTA),
		scored_distilbert = nrow(unscored_text$DISTILBERT),
		stdout = paste0(tail(read_lines(file.path(EF_DIR, 'logs', paste0(JOB_NAME, '.log'))), 100), collapse = '\n')
	)
	log_finish_in_db(db, run_id, JOB_NAME, 'sentiment-analysis', log_data)
})

## Close Connections --------------------------------------------------------
dbDisconnect(db)
message(paste0('\n\n----------- FINISHED ', format(Sys.time(), '%m/%d/%Y %I:%M %p ----------\n')))

File: modules/sentiment-analysis/sentiment-analysis-score-data.py
# Initialize ----------------------------------------------------------
# %% Imports
import os
import logging
import yaml
import torch
import pandas as pd
import polars as pl
import numpy as np
from datetime import datetime
import pytz
from sqlalchemy import create_engine
from transformers import AutoTokenizer, AutoConfig, AutoModelForSequenceClassification

# %% Constants  
JOB_NAME = 'sentiment-analysis-pull-data'
EF_DIR = os.environ['EF_DIR']
RESET_SQL = False
MAX_ROWS = 50000

# %% Logger
logging.basicConfig(
    filename = os.path.join(EF_DIR, 'logs', JOB_NAME + '.log'),
    format = '\n[%(asctime)s]\n %(message)s\n---',
    level = logging.INFO
    )
logging.getLogger().addHandler(logging.StreamHandler())


# %% Connections
with open(EF_DIR + "/model-inputs/constants.yml","r") as f: CONST = (yaml.full_load(f.read()))
db = create_engine(
    f"postgresql://{CONST['DB_USERNAME']}:{CONST['DB_PASSWORD']}@{CONST['DB_SERVER']}:5432/{CONST['DB_DATABASE']}"
    )


# Data Import ----------------------------------------------------------

# %% Create Tables
# if RESET_SQL:



# %% Boards
def get_boards():
    score_boards = pl.from_pandas(pd.read_sql(
        """
        SELECT subreddit, scrape_ups_floor FROM sentiment_analysis_reddit_boards
        WHERE score_active = TRUE
        """,
        db
    ))
    return (score_boards)

reddit = {}
(reddit['scrape_boards']) = get_boards()


# %%

# %% Pull Unscored Data 
def get_unscored_text():

    score_models = ['DICT', 'DISTILBERT', 'ROBERTA']

    def get_raw_df(x):
        raw_df = pl.from_pandas(pd.read_sql(
            f"""
			(
				SELECT
					'reddit' AS source, DATE(r1.created_dttm AT TIME ZONE 'US/Eastern') as created_dt,
					r1.id,
					CONCAT(r1.title, COALESCE(CONCAT('. ', r1.selftext), '')) AS text_part_all_text
				FROM sentiment_analysis_reddit_scrape r1
				LEFT JOIN
					(
					SELECT scrape_id--, DATE(scraped_dttm) AS scrape_date
					FROM sentiment_analysis_reddit_score
					WHERE score_model = '{x}'
					) r2
					ON r1.id = r2.scrape_id
				-- Only keep subreddits in scrape_boards directory and above score_ups_floor
				INNER JOIN sentiment_analysis_reddit_boards b
					ON r1.subreddit = b.subreddit AND b.scrape_active = TRUE AND r1.ups >= b.score_ups_floor 
				WHERE r2.scrape_id IS NULL 
					-- Rescore comments scored before 5/10/22
                    -- Uncomment  below for testing
                    -- OR DATE(r1.scraped_dttm) <= '2022-05-01'
			)
			UNION ALL
			(
				SELECT 
					'media' AS source, m1.created_dt,
					m1.id, 
					CONCAT(m1.title, COALESCE(CONCAT('. ', m1.description), '')) AS text_part_all_text
				FROM sentiment_analysis_media_scrape m1
				LEFT JOIN 
					(
					SELECT scrape_id--, DATE(scraped_dttm) AS scrape_date
					FROM sentiment_analysis_media_score
					WHERE score_model = '{x}'
					) m2
					ON m1.id = m2.scrape_id
				WHERE m2.scrape_id IS NULL
					-- Rescore comments scored before 5/10/22
					-- OR DATE(r2.scrape_date) <= '2022-05-01'
			)
            """,
            db
        ))

        return raw_df

    return {k: v for k, v in zip(score_models, [get_raw_df(x) for x in score_models])}

unscored_text = get_unscored_text()




# Analysis ----------------------------------------------------------

# %% Dictionary Analysis
def run_dict_analysis():
    BATCH_SIZE = 10000

# %% DISTILBERT
def run_distilbert():
    BATCH_SIZE = 10 #1000
    print(f'UNSCORED DISTILBERT: {len(unscored_text["DISTILBERT"])}')

    # Load pretrained
    pretrained_model = 'distilbert-base-uncased-finetuned-sst-2-english'
    tokenizer = AutoTokenizer.from_pretrained(pretrained_model)
    model = AutoModelForSequenceClassification.from_pretrained(pretrained_model)
    config = AutoConfig.from_pretrained(pretrained_model)

    batches =\
        unscored_text["DISTILBERT"]\
        .head(MAX_ROWS)\
        .sort(['source', 'created_dt'], reverse = [True, True])\
        .pipe(lambda x: x.with_column(pl.Series(list(range(1, len(x) + 1))).alias('split')))\
        .with_column((pl.col('split')/BATCH_SIZE).ceil())\
        .partition_by(['source', 'split'])

    def get_scores_and_send_to_sql(batch):
        encoded_tokens = tokenizer(
            batch['text_part_all_text'].to_list(),
            padding = True,
            truncation = True,
            max_length = 512,
            return_tensors = "pt"
        )
        predictions = model(**encoded_tokens).logits
        predictions_normalized = torch.nn.Softmax(dim=-1)(predictions).cpu().detach().numpy()
        sql_data =\
            pl.concat([
                pl.from_dict({'score': list(config.id2label.values()), 'score_conf': x.tolist(), 'idx': [i] * 2})
                for i, x in enumerate(predictions_normalized)
            ])\
            .with_columns([
                pl.when(pl.col('score') == 'POSITIVE').then('p').otherwise('n').alias('score'),
                pl.Series(list(np.repeat(batch['id'].to_list(), 2))).alias('scrape_id'),
                pl.lit('all_text').alias('text_part'),
                pl.lit('DISTILBERT').alias('score_model'),
                pl.lit(datetime.now(pytz.timezone('US/Eastern')).strftime('%Y-%m-%d %H:%M:%S %z')).alias('scored_dttm'),
                ])\
            [['scrape_id', 'text_part', 'score_model', 'score', 'score_conf', 'scored_dttm']]

        ## Create insert_sql query            

    get_scores_and_send_to_sql(batches[1])
    [get_scores_and_send_to_sql(batch) for batch in batches]

run_distilbert()


# %% Pull Unscored Data
pd_df = pd.DataFrame(db.execute(
    "SELECT scrape_id, text_part, score_model, score, score_conf FROM sentiment_analysis_reddit_score LIMIT 10"
    ).fetchall())


# %% Polars Test
df = pl.from_dicts(pd_df.to_dict('records'))
(df
    .filter(pl.col('score_model') == 'ROBERTA')
    .groupby('score')
    .agg([
        pl.count().alias('count')
    ])
)


# %% Load Pretrained
pretrained_model = 'siebert/sentiment-roberta-large-english'
tokenizer = AutoTokenizer.from_pretrained(pretrained_model)
model = AutoModelForSequenceClassification.from_pretrained(pretrained_model)
config = AutoConfig.from_pretrained(pretrained_model)

# %% Tokenize & Predict
encoded_tokens = tokenizer(
    ["Test Str1.", "Test string 2"],
    padding = True,
    truncation = True,
    max_length = 512,
    return_tensors = "pt"
)
predictions = model(**encoded_tokens).logits
predictions_normalized = torch.nn.Softmax(dim=-1)(predictions).cpu().detach().numpy()
pl.concat(
    [
        pl.from_dict({'labels': list(config.id2label.values()), 'score': x.tolist(), 'idx': [i] * 2})
        for i, x in enumerate(predictions_normalized)
    ]
)

# pd.concat([pd.DataFrame({'labels': config.id2label.values(), 'score': x, 'idx': i}) for i, x in enumerate(predictions_normalized)])

# %% Import Data
tokenizer = RobertaTokenizer.from_pretrained('roberta-base')

raw_data = pd.read_csv('test.csv').to_dict('records')
tokenized_data = []
for x in raw_data:
    # Leave off 2 tokens, since we need to append one to start & end
    cut_text = x['text'][:512-2]
    tokens = tokenizer.tokenize(cut_text)
    indexed_tokens = tokenizer.encode(cut_text)
    tokenized_data.append(tokens)
    


# # %% Import Dataset

# # Load One at a Time
# class TextDataset(Dataset):
#     def __init__(self, filepath, transform=None):
#         self.data = pd.read_csv(filepath)
#         self.max_len = 512
#         self.transform = transform

#     def __len__(self):
#         return len(self.filepath)

#     def __getitem__(self, idx):
#         if torch.is_tensor(idx):
#             idx = idx.tolist()
        
#         sample = self.data.iloc[idx].to_dict().text
        
#         if self.transform:
#             sample = self.transform(sample)

#         return sample

# text_dataset = TextDataset('test.csv')


# # %% Import Tokenizer
# tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
# logging.info(f'Vocab Size\n {tokenizer.vocab_size}')
# logging.info(f'Max Token Sizes\n {tokenizer.max_model_input_sizes}')

# tokens = tokenizer.tokenize('Hello WORLD how ARE yoU?')
# logging.info(tokens)

# indexes = tokenizer.convert_tokens_to_ids(tokens)
# logging.info(indexes)

# def tokenize_and_cut(sentence):
#     tokens = tokenizer.tokenize(sentence) 
#     tokens = tokens[:512-2] # Leave off 2 tokens, since we need to append one to start & end
#     return tokens


# # %% 
# model = RobertaModel.from_pretrained("roberta-base")

# inputs = tokenizer("Hello, I hate my dog", return_tensors="pt")
# outputs = model(**inputs)

# # %%
# torch.utils.data
# # %%
